src/writer.js

```diff
@@ -23,53 +24,71 @@ class Writer {
     if (trace.started.length === trace.finished.length) {
       const formattedTrace = trace.finished.map(format)
 
-      log.debug(() => `Encoding trace: ${JSON.stringify(formattedTrace)}`)
-
-      const buffer = encode(formattedTrace)
```

Comment:
This change so that we add the **encoded** trace to the buffer, rather than just the **formatted**, could be avoided if there's a good reason why this approach is better. But then we'd need to change the semantics away from `beforeFlush`, which I picked for consistency with the Ruby implementation of this feature.

Reply:
There were 2 main reasons for this:

* Encoding is the most expensive operation of the tracer. By encoding it early, the load is distributed. This avoids locking given the asynchronous nature of Node.
* We will eventually implement a check on the total buffer size to flush right away if it gets to large. Since this should be triggered whenever adding a trace, we have to encode it to know the buffer size.

I think what we do in [Python](http://pypi.datadoghq.com/trace/docs/advanced_usage.html#trace-filtering) is closer to what would work in Node.

Basically there could be a `processors` option that takes either an array or a single element. Each element could be either a class with a `process` method, or a function with the same signature as the method. We can revisit the wording later but basically that would be the idea. The processor(s) would then be called on every append whenever the trace is finished.